diff --git a/monitoring/histogram.cc b/monitoring/histogram.cc
index c6ef856b0..a7d4de3f8 100644
--- a/monitoring/histogram.cc
+++ b/monitoring/histogram.cc
@@ -186,8 +186,8 @@ std::string HistogramStat::ToString() const {
   r.append(buf);
   snprintf(buf, sizeof(buf),
            "Percentiles: "
-           "P50: %.2f P75: %.2f P99: %.2f P99.9: %.2f P99.99: %.2f\n",
-           Percentile(50), Percentile(75), Percentile(99), Percentile(99.9),
+           "P50: %.2f P75: %.2f P90: %.2f P95: %.2f P99: %.2f P99.9: %.2f P99.99: %.2f\n",
+            Percentile(50), Percentile(75), Percentile(90), Percentile(95), Percentile(99), Percentile(99.9),
            Percentile(99.99));
   r.append(buf);
   r.append("------------------------------------------------------\n");
